thunder/core/transforms.py

```diff
@@ -1357,6 +1357,27 @@ def _embedding_prim_grad(
 
 register_grad(pids.EMBEDDING, _embedding_prim_grad)
 
+
+def _maximum_grad(a: TensorProxy, b: TensorProxy, /):
+    fwd = prims.maximum(a, b)
+
+    g = get_grad(fwd)
+
+    a_grad = prims.where(a == b, g / 2, 0.0)
+    b_grad = prims.where(a == b, g / 2, 0.0)
+    a_grad = prims.where(a > b, g, a_grad)
+    b_grad = prims.where(b > a, g, b_grad)
```

Comment:
Took me a bit to realize that what we are computing is a subgradient :)

Reply:
nit: Maybe we should isolate `g / 2`? Or is it optimized out from the trace?

Reply:
We just have one `g/2` in the final trace.

```python
import torch
import thunder

def foo(*args):
    return torch.max(*args)

jfoo = thunder.jit(foo)

x = torch.randn(3, 3, requires_grad=True)

def _test(*args):
    expected = foo(*args)
    actual = jfoo(*args)
    torch.testing.assert_close(actual, expected)

# _test(x)
# _test(x, 0)
# _test(x, 0, True)
_test(x, x + 1)

print(thunder.last_backward_traces(jfoo)[-1])

```

Backward Trace
```python
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, _, = saved_for_backward
  clear_mutable_collection(saved_for_backward)
  del saved_for_backward
  t1, = cotangents
  clear_mutable_collection(cotangents)
  del cotangents
  t_0, t_1, = C0
  clear_mutable_collection(C0)
  del C0
  t12 = torch.true_divide(t1, 2)  # t12: "cpu f32[3, 3]"
    # t12 = ltorch.true_divide(t1, 2)  # t12: "cpu f32[3, 3]"
      # _ = prims.convert_element_type(2, float)
      # t12 = prims.div(t1, 2.0)  # t12: "cpu f32[3, 3]"
  t11 = torch.eq(t_0, t_1)  # t11: "cpu b8[3, 3]"
    # t11 = ltorch.eq(t_0, t_1)  # t11: "cpu b8[3, 3]"
      # t11 = prims.eq(t_0, t_1)  # t11: "cpu b8[3, 3]"
  t16 = torch.gt(t_1, t_0)  # t16: "cpu b8[3, 3]"
    # t16 = ltorch.gt(t_1, t_0)  # t16: "cpu b8[3, 3]"
      # t16 = prims.gt(t_1, t_0)  # t16: "cpu b8[3, 3]"
  t14 = torch.gt(t_0, t_1)  # t14: "cpu b8[3, 3]"
    # t14 = ltorch.gt(t_0, t_1)  # t14: "cpu b8[3, 3]"
      # t14 = prims.gt(t_0, t_1)  # t14: "cpu b8[3, 3]"
  del t_0, t_1
  t13 = torch.where(t11, t12, 0.0)  # t13: "cpu f32[3, 3]"
    # t13 = ltorch.where(t11, t12, 0.0)  # t13: "cpu f32[3, 3]"
      # t13 = prims.where(t11, t12, 0.0)  # t13: "cpu f32[3, 3]"
  del t11, t12
  t17 = torch.where(t16, t1, t13)  # t17: "cpu f32[3, 3]"
    # t17 = ltorch.where(t16, t1, t13)  # t17: "cpu f32[3, 3]"
      # t17 = prims.where(t16, t1, t13)  # t17: "cpu f32[3, 3]"
  del t16
  t15 = torch.where(t14, t1, t13)  # t15: "cpu f32[3, 3]"
    # t15 = ltorch.where(t14, t1, t13)  # t15: "cpu f32[3, 3]"
      # t15 = prims.where(t14, t1, t13)  # t15: "cpu f32[3, 3]"
  del t14, t1, t13
  return (t15, t17)
```

Reply:
I think it might be prettier to not rely on CSE for it, but seems OK for now.